app/handlers/handlers_groups.py

```diff
@@ -0,0 +1,85 @@
+from datetime import timedelta
+
+import kopf
+
+from app.api import TwingateAPIClient
+from app.api.exceptions import GraphQLMutationError
+from app.crds import GroupStatusUserID, TwingateGroupCRD
+from app.handlers import fail, success
+
+
+@kopf.on.resume("twingategroup")
+@kopf.on.create("twingategroup")
+@kopf.on.update("twingategroup")
+def twingate_group_create_update(body, spec, logger, memo, patch, **kwargs):
+    logger.info("twingate_group_reconciler: %s", spec)
+    settings = memo.twingate_settings
+    client = TwingateAPIClient(settings)
+
+    if diff := kwargs.get("diff"):
+        logger.info("Diff: %s", diff)
+        # If ID changed from `None to value we just created it no need to update
+        # diff is `(('add', ('spec', 'id'), None, 'R3JvdXA6MjAxNjc4OQ=='),)`
+        if (
+            len(diff) == 1
+            and diff[0][0] == "add"
+            and diff[0][1] == ("spec", "id")
+            and diff[0][2] is None
+        ):
+            return success()
+
+    crd = TwingateGroupCRD(**body)
+    user_ids = crd.status.user_ids if crd.status else []
+    if not crd.is_user_ids_cache_valid:
+        logger.info("User IDs cache is invalid, fetching from Twingate")
+        kopf.info(body, reason="User IDs cache is invalid, fetching from Twingate")
+
+        email_to_id = client.get_ids_for_emails(crd.spec.members_email)
+        user_ids = [
+            GroupStatusUserID(id=id, email=email) for email, id in email_to_id.items()
+        ]
+        user_ids.extend([GroupStatusUserID(id=uid) for uid in crd.spec.members_ids])
+        patch.status["user_ids"] = [ui.model_dump() for ui in user_ids]
+        patch.status["user_ids_hash"] = crd.spec.members_hash
```

Comment:
I think there is an edge case in caching the email lookup. Let's say the user's email is `foo@gmail.com` and the group k8s spec is 
```yaml
spec:
  name: Example Group
  members:
    - "foo@gmail.com"
```
If that user's email is changed to `bar@gmail.com`, the user still belongs to this `Example Group`. This doesn't happen very often though. Maybe we should have a cache expiration like 1 day or 1 week? 🤔